prost-derive/src/lib.rs

```diff
@@ -251,14 +251,283 @@ fn try_message(input: TokenStream) -> Result<TokenStream, Error> {
+                }
+            );
+
+            outer_path.path.segments.push(option_segment);
+            syn::Type::Path(outer_path)
+        },
+        _ => {
+            bail!("oneof field is not of type Path")
+        },
+    };
+    Ok(field)
 }
 
-#[proc_macro_derive(Message, attributes(prost))]
-pub fn message(input: TokenStream) -> TokenStream {
-    try_message(input).unwrap()
+/// Applies `convert_oneof_field_option_to_openvariant` to all fields of a struct
+fn convert_oneof_options_to_openenum(
+    data: syn::Data
+) -> Result<syn::Data, Error> {
+
+    let mut variant_data = match data {
+        Data::Struct(variant_data) => variant_data,
+        Data::Enum(..) => bail!("open_oneof can not be derived for an enum"),
+        Data::Union(..) => bail!("open_oneof can not be derived for a union"),
+    };
+
+    variant_data.fields = match variant_data.fields {
+        Fields::Named(fields) => {
+            Fields::Named(FieldsNamed{
+                named: fields.named.into_iter().map(|field| {
+                    add_openvariant_to_oneof_field(field)
+                }).try_collect()?,
+                ..fields
+            })
+        }
+        Fields::Unnamed(fields) => {
+            Fields::Unnamed(FieldsUnnamed{
+                unnamed: fields.unnamed.into_iter().map(|field| {
+                    add_openvariant_to_oneof_field(field)
+                }).try_collect()?,
+                ..fields
+            })
+        }
+        fields @ Fields::Unit => fields,
+
+    };
+    Ok(Data::Struct(variant_data))
 }
 
+#[proc_macro_attribute]
+pub fn open_oneof(_attr: TokenStream, item: TokenStream) -> TokenStream {
+    let methods = try_message(item.clone()).expect("failed to compute methods");
+
+    let mut derive_input = syn::parse_macro_input!(item as DeriveInput);
+
+    let mut prost_messages_removed = 0;
+    let new_attrs = derive_input
+        .attrs
+        .into_iter()
+        .map(|mut attr| {
+            if !&attr.meta.path().is_ident("derive") {
+                return attr;
+            }
+
+            let syn::Meta::List(ref mut meta_list) = &mut attr.meta else {
+                return attr;
+            };
+
+            let (tokens, incr) = strip_prost_messages_from_procmacro_attrs(std::mem::take(&mut meta_list.tokens));
+            meta_list.tokens = tokens;
+            prost_messages_removed = prost_messages_removed + incr;
+
+            attr
+        })
+        .collect();
+
+    if prost_messages_removed != 1 {
+        panic!("`open_oneof` procmacro should only be attached above a prost::Message")
+    }
+    derive_input.attrs = new_attrs;
+
+    derive_input.data = convert_oneof_options_to_openenum(derive_input.data).expect("failed to convert_oneof_options_to_openenum");
+    derive_input.data = strip_field_attrs(derive_input.data).expect("failed to strip_field_attrs");
+
+    let x = quote! {
+        #methods
+
+        #derive_input
+    };
```

Comment:
This is where all is assembled. `methods` are the patched methods generated by `try_message`. `derive_input` is the patched struct